### DISCUSSION TITLE
FV variable gradient in FVElemental kernel

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#22484] [Q&A General] [answered]
I need to write an FV kernel that inherits from FVElementalKernel but includes gradients of FV variables. I understand that I cannot just use coupledGradient as I would for the usual FE variables. Is there a "ready-made" way to access or compute (element-averages of)  gradients of FV variables using the procedures outlined in "Reconstruction" on https://mooseframework.inl.gov/moose/finite_volumes/fv_design.html, or do I have to figure out how to compute the element-average gradient?

Thanks.

**Comment by GiudGiud** [comment]
Hello

Did you try the functor element gradient?
https://mooseframework.inl.gov/docs/doxygen/moose/classMoose_1_1FunctorBase.html#a9af8d394649b42b3c4ca9c4c7ba07e42

Guillaume

**Comment by heinono1** [comment]
Hi,

Thanks. Sorry for being a bit slow on the uptake, but do you know of an
example of usage?

On Mon, Oct 24, 2022 at 4:22 PM Guillaume Giudicelli <
***@***.***> wrote:

> Hello
>
> Did you try the functor element gradient?
>
> https://mooseframework.inl.gov/docs/doxygen/moose/classMoose_1_1FunctorBase.html#a9af8d394649b42b3c4ca9c4c7ba07e42
>
> Guillaume
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/22484#discussioncomment-3954959>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEFYT7YCH4PKVOYQ2J33WE34YXANCNFSM6AAAAAARNKQ5U4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
I guess I dont
I thought this one had
https://github.com/idaholab/moose/blob/next/modules/navier_stokes/src/fvkernels/INSFVMixingLengthReynoldsStress.C

but it uses another API, adGradSln, which should work too

**Reply by lindsayad** [reply]
An example of what @GiudGiud suggested (which is a great option), is [here](https://github.com/idaholab/moose/blob/next/test/src/fvkernels/FVElementalAdvection.C#L55)

**Comment by heinono1** [comment]
Thanks. I'll give it a try.

On Mon, Oct 24, 2022 at 5:37 PM Guillaume Giudicelli <
***@***.***> wrote:

> I guess I dont
> I thought this one had
>
> https://github.com/idaholab/moose/blob/next/modules/navier_stokes/src/fvkernels/INSFVMixingLengthReynoldsStress.C
>
> but it uses another API, adGradSln, which should work too
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/22484#discussioncomment-3955284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEFYIDDW2WPSD5WUYRSDWE4FSXANCNFSM6AAAAAARNKQ5U4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Comment by lindsayad** [comment, marked as answer]
We're trying to encourage our posters to use threads in their responses to answers. You can't use `coupleGradient`, but you should be able to use `adCoupledGradient` as you would for an AD kernel in FE

**Reply by heinono1** [reply]
Thanks, Alex. Sorry about breaking the thread.

**Reply by lindsayad** [reply]
No problem at all!

**Reply by heinono1** [reply]
@lindsayad so does adCoupledGradient work in the FV system so I can use it directly to calculate an element-average gradient (FVElementalKernel) without any other extra steps (eg getting element faces etc etc)?

**Reply by lindsayad** [reply]
Correct. For a finite volume variable, `adCoupledGradient` returns a cell-average gradient computed via Green-Gauss

**Reply by heinono1** [reply]
That is awesome! Thank you, MOOSE team!
